### PR TITLE
Feat/ neo4j date attr type

### PR DESCRIPTION
**Pull request by CasperCDR**
Added _Neo4jDate in mutations __init__.py which can take a python datetime or a value representing the year.

### CODE REVIEW

File: trompace/exceptions.py [hunk 1]
```diff
@@ -19,7 +19,7 @@ def __init__(self, object):
         super().__init__("{} not found in configuration file".format(object))
 
 
-class VaueNotFound(Exception):
```
Comment (alastair) [inline trompace/exceptions.py]: could you make a separate PR for this so that we can get the fixed merged, and continue with the discussion on the date stuff?

File: trompace/mutations/person.py [hunk 2]
```diff
@@ -90,7 +90,7 @@ def mutation_create_artist(artist_name: str, publisher: str, contributor: str, c
     if coverage:
         args["coverage"] = coverage
     if date:
-        args["date"] = date
+        args["date"] = _Neo4jDate(date)
```
Comment (alastair) [inline trompace/mutations/person.py]: Instead of creating this object here, what about we specify that the type of this argument should be a `datetime.date` or `datetime.datetime` object? This way, we can render the formatted date in the `mutations.make_parameters` method.

File: trompace/mutations/__init__.py [hunk 3]
```diff
@@ -13,6 +14,20 @@ def __repr__(self):
...
+class _Neo4jDate(StringConstant):
+    """The _Neo4jDate is used for Date values. It will be added as
+    StringConstant in the GraphQL. The date will be formatted
+    as: { year:[int] [month: [int] day: [int]] }
+
+    The constructor argument can be a datetime or a year."""
+
+    def __init__(self, value):
+        if isinstance(value, datetime):
+            self.value = "{{ year: {0} month: {1} day: {2} }}".format(value.year, value.month, value.day)
```
Comment (alastair) [inline trompace/mutations/__init__.py]: we need to be careful as to what values we want to support here - we could say that a birth/death date could be represented as a `datetime.date` (not datetime). In MusicBrainz, this value can be yyyy, or yyyy-mm or yyyy-mm-dd. I don't know if the python type can represent just a yyyy or yyyy-mm.

In the case that we want to represent a `created` field for an object, this should also contain the hour, minute, second values, We should also make sure that this kind of conversion properly handles this case (and add tests).

We had some discussion about how to represent dates for which we didn't have exact values (e.g. "around 1710", "born on the 27th or 28th"). We haven't talked about this in much more detail, so I think that we can skip it for now, but we should think of a nice way to represent it in this client too.

File: trompa_ce_client.egg-info/PKG-INFO [hunk 4]
```diff
@@ -0,0 +1,46 @@
+Metadata-Version: 2.1
```
Comment (alastair) [inline trompa_ce_client.egg-info/PKG-INFO]: looks like you mistakenly included this egg-info directory and the egg file (in dist/). It'd be good to rebase to remove these.

File: trompace/mutations/__init__.py (hunk 3 above)
Comment (CasperCDR) [reply trompace/mutations/__init__.py]: I agree that the `datetime.date` type has the right precision for the field. A python date type must contain a valid date, so with month and day included. Regarding support for year-month support we can consider checking for a list type as value. The values should contain [year, month, day] where day and month are optional.

The uncertain dates can not be represented by the current _Neo4jDate type. In Schema.org the date type is defined by the ISO 8601 standards where in the basic rules there is also no support, but in the [extensions](https://www.iso.org/standard/70908.html) there is.

File: trompace/mutations/person.py (hunk 2 above)
Comment (CasperCDR) [reply trompace/mutations/person.py]: The date type is connected to the attributes of the types in the dataset, not to the type we try to post as parameter. So I don't think your proposal is a good idea. The second argument is that a date or datetime type does not support only a year or year+month formats.

File: trompa_ce_client.egg-info/PKG-INFO (hunk 4 above)
Comment (CasperCDR) [reply trompa_ce_client.egg-info/PKG-INFO]: Since I split the changes in two branches the .gitignore got replaced by the original one. Fixed it in the current state.

File: trompace/exceptions.py (hunk 1 above)
Comment (CasperCDR) [reply trompace/exceptions.py]: Done, see [pull request 23](https://github.com/trompamusic/trompa-ce-client/pull/23)

File: trompace/mutations/__init__.py (hunk 3 above)
Comment (alastair) [reply trompace/mutations/__init__.py]: because the CE is currently using the _Neo4jDate type, let's not focus on uncertain dates. We can come back to this in the future if we find a clear need for it.

Comment (CasperCDR) [reply trompace/mutations/__init__.py]: Added support for the suggested year+month support. Also included tests for the _Neo4jDate type in test_mutations.py. Please verify the solution.

**Comment by alastair** [comment]
This looks good in general. I tried some queries against the CE and got these results:
```
mutation {
  CreatePerson (
    title: "name"
    contributor:"alastair"
    format: "text/html"
    source:"foo"
    creator:"alastiar"
    name: "aname"
    date: {year: 2000}
  ){
    identifier
  }
}
```

```
query {
  Person(identifier: "b16777d3-e9f3-49be-8fc8-098ad30f66dc") {
    date {
      formatted
    }
  }
```

gives a result 

```
{
  "data": {
    "Person": [
      {
        "date": {
          "formatted": "2000-01-01"
        }
      }
    ]
  }
}
```

In my view, this is incorrect. Setting just a year shouldn't set the month and day to 01, because these have distinct meanings. I'll open an issue in ce-api to discuss this.

Do you have a specific requirement to support partial dates? If all of your data has a full date, perhaps we could revert your most recent change and _only_ allow datetime.date objects here, so that we are sure that we will always get a complete date object. What do you think?
}

**Comment by CasperCDR** [comment]
@alastair: In reply to your last comment and test for the date values in the dataset I agree that the date should not be set with values not supplied. The date type could be made more strict and only support valid dates, so a rollback to support only python date values. 
To support only years (or year+month as you mentioned from the supported formats in MusicBrainz) for birthDate and deathDate is desired because at Muziekweb we only store years in our current datamodel, no exact dates.

**Comment by alastair** [comment]
I've opened a ticket for this in the CE. Let's assume that we fix this eventually.
For now, could you please update the types and docstring in the create_ and update_ methods where needed, then we'll merge this.

**Comment by CasperCDR** [comment]
Thanks for opening the ticket. The docstrings are updated.
